File: src/services/cart.service.ts
import { CartRepository } from "@src/adapters/persistence/repository/cart.repository";
import { PickRepository } from "@src/adapters/persistence/repository/pick.repository";
import { CartDomain } from "@src/domain/cart.domain";
import { ErrorDomain } from "@src/domain/error.domain";
import { PageInfo } from "@src/dto/pageInfo.dto";
import mongoose from "mongoose";

const cartRepository = new CartRepository();
const pickRepository = new PickRepository();
const isValidObjectId = (id: string): boolean => mongoose.isValidObjectId(id);

export class CartService {
  async getMyCarts(
    userId: string,
    page: number,
    size: number
  ): Promise<{
    carts: {
      id: string;
      name: string;
      picksNum: number;
      picksImages: string[];
    }[];
    page: PageInfo;
  }> {
    console.log(userId);
    if (page < 1) {
      throw new ErrorDomain("page must be larger than 0", 400);
    }

    if (size < 1) {
      throw new ErrorDomain("size must be larger than 0", 400);
    }

    const cartDomains = await cartRepository.findByUserId(userId, page, size);

    if (!cartDomains) {
      throw new ErrorDomain("Cart not found by userId", 404);
    }
    const resCart = await Promise.all(
      cartDomains.carts.map(async (cart) => {
        // 각 pick_id에 대해 pick 정보를 가져옵니다.
        const pickIds = cart.pickItemIds;
        const picks = (
          await pickRepository.findByPickIds(pickIds, 1, cart.pickNum)
        ).picks;

        // pick 이미지들을 모아 하나의 배열로 만듭니다.
        const pickImages = picks.flatMap((pick) => pick.image).slice(0, 3); // 최대 3개의 이미지만 포함

        // 새로운 카트 객체를 만듭니다.
        return {
          id: cart.id!,
          name: cart.name,
          picksNum: pickIds.length,
          picksImages: pickImages,
        };
      })
    );

    return { carts: resCart, page: cartDomains.page };
  }

  async deleteCart(userId: string, cartId: string): Promise<string> {
    const cartInfo = await cartRepository.findById(cartId);
    if (cartInfo.userId != userId) {
      throw new ErrorDomain("not cart owner", 400);
    }
    const deletedCartId = await cartRepository.delete(cartId);
    if (!deletedCartId) {
      throw new ErrorDomain("Cart not found", 404);
    }
    return deletedCartId;
  }

  async createCart(
    userId: string,
    name: string,
    pickIds: string[]
  ): Promise<CartDomain> {
    if (!name) {
      throw new ErrorDomain("'name' is required", 400);
    }
    const cart = await cartRepository.create(userId, name, pickIds);
    if (cart) {
      return cart;
    }
    throw new ErrorDomain("Cart not created", 404);
  }

  async movePicks(
    userId: string,
    pickIds: string[],
    sourceCartId: string,
    destinationCartId: string,
    isDeleteFromOrigin: boolean
  ): Promise<{
    source: { cartId: string; pickIds: string[] };
    destination: { cartId: string; pickIds: string[] };
  }> {
    if (isDeleteFromOrigin == null) {
      throw new ErrorDomain("'isDeleteFromOrigin' filed is required", 400);
    }
    if (!destinationCartId) {
      throw new ErrorDomain("'destinationCartId' filed is required", 400);
    }
    if (isDeleteFromOrigin == false) {
      const movedPickIds: string[] | null = await cartRepository.addToCart(
        pickIds,
        destinationCartId
      );
      if (!movedPickIds) {
        throw new ErrorDomain("already exist in cart", 400);
      }
      if (pickIds.length == 0) {
        throw new ErrorDomain("pickIds is required", 400);
      }
      const picksInfo = await pickRepository.findByPickIds(
        pickIds,
        1,
        pickIds.length
      );
      picksInfo.picks.forEach((el) => {
        if (el.userId != userId) {
          throw new ErrorDomain("not pick owner", 400);
        }
      });
      return {
        source: {
          cartId: sourceCartId ? sourceCartId : "__all__",
          pickIds: movedPickIds,
        },
        destination: { cartId: destinationCartId, pickIds: movedPickIds },
      };
    }
    if (isDeleteFromOrigin == true) {
      const movedPickIds: string[] | null = await cartRepository.moveCart(
        pickIds,
        sourceCartId,
        destinationCartId
      );
      if (!movedPickIds) {
        throw new ErrorDomain("can not moved picks", 500);
      }
      return {
        source: {
          cartId: sourceCartId ? sourceCartId : "__all__",
          pickIds: [],
        },
        destination: { cartId: destinationCartId, pickIds: movedPickIds },
      };
    }
    throw new ErrorDomain("can not moved picks with server error", 500);
  }
}

File: src/main/java/kr/okku/server/service/CartService.java
package kr.okku.server.service;

import kr.okku.server.adapters.persistence.CartPersistenceAdapter;
import kr.okku.server.adapters.persistence.PickPersistenceAdapter;
import kr.okku.server.domain.CartDomain;
import kr.okku.server.domain.Log.TraceId;
import kr.okku.server.dto.controller.cart.*;
import kr.okku.server.exception.ErrorCode;
import kr.okku.server.exception.ErrorDomain;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

@Service
public class CartService {

    private final CartPersistenceAdapter cartPersistenceAdapter;

    private final PickPersistenceAdapter pickPersistenceAdapter;

    @Autowired
    public CartService(CartPersistenceAdapter cartPersistenceAdapter, PickPersistenceAdapter pickPersistenceAdapter) {
        this.cartPersistenceAdapter = cartPersistenceAdapter;
        this.pickPersistenceAdapter = pickPersistenceAdapter;

    }

    public MyCartsResponseDto getMyCarts(TraceId traceId,String userId) {
        List<CartDomain> cartDomains = cartPersistenceAdapter.findByUserId(userId);

        if (cartDomains == null) {
            throw new ErrorDomain(ErrorCode.CART_NOT_EXIST,traceId);
        }

        var resCart = cartDomains.stream().map(cart -> {
            List<String> pickIds = cart.getPickItemIds();
                var picks = pickPersistenceAdapter.findByIdIn(pickIds);

            List<String> pickImages = picks.stream()
                    .map(pick -> pick.getImage())
                    .limit(3)
                    .collect(Collectors.toList());

            return new CartDto(
                    cart.getId(),
                    cart.getName(),
                    pickIds.size(),
                    pickImages
            );
        }).collect(Collectors.toList());

        MyCartsResponseDto responseDto =  new MyCartsResponseDto();
        responseDto.setCarts(resCart);
        return responseDto;
    }



    @Transactional
    public String deleteCart(TraceId traceId, String userId, String cartId) {
        CartDomain cartInfo = cartPersistenceAdapter.findById(cartId)
                .orElseThrow(() -> new ErrorDomain(ErrorCode.CART_NOT_EXIST,traceId));
        if (!cartInfo.getUserId().equals(userId)) {
            throw new ErrorDomain(ErrorCode.NOT_CART_OWNER,traceId);
        }
        cartPersistenceAdapter.deleteById(cartId);
        return cartId;
    }

    @Transactional
    public CartDomain createCart(TraceId traceId, String userId, CreateCartRequestDto requestDto) {
        String name = requestDto.getName();
        List<String> pickIds = requestDto.getPickIds();

        if (name == null || name.isEmpty()) {
            throw new ErrorDomain(ErrorCode.NAME_IS_EMPTY,traceId);
        }
        Integer size = 0;
        if(pickIds==null){
            pickIds = new ArrayList<>();
        }
        CartDomain cart =  CartDomain.builder()
                .pickItemIds(pickIds)
                .userId(userId)
                .name(name)
                .pickNum(size)
                .build();
        List<CartDomain> userCarts = cartPersistenceAdapter.findByUserId(userId);
        int maxOrderIndex = userCarts.stream()
                .mapToInt(CartDomain::getOrderIndex)
                .max()
                .orElse(0); // Default to 0 if no carts exist

        cart.setOrderIndex(maxOrderIndex + 1);
        CartDomain savedCart = cartPersistenceAdapter.save(cart);
        if (savedCart != null) {
            return savedCart;
        }
        throw new ErrorDomain(ErrorCode.DATABASE_ERROR_WITH_SAVE_CART,traceId);
    }


    @Transactional
    public MyCartsResponseDto updateCarts(TraceId traceId,String userId, UpdateCartsRequestDto requestDto) {
        List<CartDomain> userCarts = cartPersistenceAdapter.findByUserId(userId);

        Set<String> updateCartIds = requestDto.getCarts().stream()
                .map(UpdateCartRequestDto::getId)
                .collect(Collectors.toSet());

        List<CartDomain> deletedCarts = userCarts.stream()
                .filter(cart -> !updateCartIds.contains(cart.getId()))
                .collect(Collectors.toList());

        Set<String> originCartIds = userCarts.stream()
                .map(CartDomain::getId)
                .collect(Collectors.toSet());

        List<UpdateCartRequestDto> createdCarts = requestDto.getCarts().stream()
                .filter(cart -> !originCartIds.contains(cart.getId()))
                .collect(Collectors.toList());

        deletedCarts.forEach((cartDomain -> {
            this.deleteCart(traceId,userId,cartDomain.getId());
        }));

        createdCarts.forEach((cartDomain -> {
            CreateCartRequestDto cartRequestDto = new CreateCartRequestDto();
            cartRequestDto.setName(cartDomain.getName());
            this.createCart(traceId,userId,cartRequestDto);
        }));

        List<CartDomain> updatedUserCarts = cartPersistenceAdapter.findByUserId(userId);
        for (int i = 0; i < requestDto.getCarts().size(); i++) {
            UpdateCartRequestDto cartRequestDto = requestDto.getCarts().get(i);
            int finalI = i;
            updatedUserCarts.stream()
                    .filter(cart -> cart.getId().equals(cartRequestDto.getId()))
                    .findFirst()
                    .ifPresent(cart -> {
                        cart.setOrderIndex(finalI + 1);
                        cart.setName(cartRequestDto.getName());
                        cartPersistenceAdapter.save(cart);
                    });
        }

        return getMyCarts(traceId,userId);
    }
}
